awx_collection/plugins/modules/tower_export.py

```diff
@@ -0,0 +1,155 @@
+      description:
+        - organization name to export
+      default: ''
+      type: str
+    user:
+      description:
+        - user name to export
+      default: ''
+      type: str
+    team:
+      description:
+        - team name to export
+      default: ''
+      type: str
+    credential_type:
+      description:
+        - credential type name to export
+      default: ''
+      type: str
+    credential:
+      description:
+        - credential name to export
+      default: ''
+      type: str
+    notification_template:
+      description:
+        - notification template name to export
+      default: ''
+      type: str
+    inventory_script:
+      description:
+        - inventory script name to export
+      default: ''
+      type: str
+    inventory:
+      description:
+        - inventory name to export
+      default: ''
+      type: str
+    project:
+      description:
+        - project name to export
+      default: ''
+      type: str
+    job_template:
+      description:
+        - job template name to export
+      default: ''
+      type: str
+    workflow:
+      description:
+        - workflow name to export
+      default: ''
+      type: str
+requirements:
+  - "awxkit >= 9.3.0"
+notes:
+  - Specifying a name of "all" for any asset type will export all items of that asset type.
+extends_documentation_fragment: awx.awx.auth
+'''
+
+EXAMPLES = '''
+- name: Export all tower assets
+  tower_export:
+    all: True
+- name: Export all inventories
+  tower_export:
+    inventory: 'all'
+- name: Export a job template named "My Template" and all Credentials
+  tower_export:
+    job_template: "My Template"
+    credential: 'all'
+'''
+
+from os import environ
+
+from ..module_utils.tower_awxkit import TowerAWXKitModule
+
+try:
+    from awxkit.api.pages.api import EXPORTABLE_RESOURCES
+    HAS_EXPORTABLE_RESOURCES=True
+except ImportError:
+    HAS_EXPORTABLE_RESOURCES=False
+
+
+def main():
+    argument_spec = dict(
+        all=dict(type='bool', default=False),
+    )
+
+    # We are not going to raise an error here because the __init__ method of TowerAWXKitModule will do that for us
+    if HAS_EXPORTABLE_RESOURCES:
+        for resource in EXPORTABLE_RESOURCES:
+            argument_spec[resource] = dict()
+
+    module = TowerAWXKitModule(argument_spec=argument_spec)
+
+    if not HAS_EXPORTABLE_RESOURCES:
+        module.fail_json(msg="Your version of awxkit does not have import/export")
+
+    # The export process will never change a Tower system
+    module.json_output['changed'] = False
+
+    # The exporter code currently works like the following:
+    #   Empty list == all assets of that type
+    #   string = just one asset of that type (by name)
+    #   None = skip asset type
+    # Here we are going to setup a dict of values to export
+    export_args = {}
+    for resource in EXPORTABLE_RESOURCES:
+        if module.params.get('all') or module.params.get(resource) == 'all':
```

Comment:
Seems like we should add some really simple validation for the case that 

```python
module.params.get('all') and module.params.get(resource) not in (None, 'all')
```

In that case, you're getting conflicting instructions from the user. The "all" parameter dictates that you should export all credentials, but the credential param says to export just "my-cred", for example.

Reply:
Should we consider moving away from the use of `'all'` flags?  The actual semantics of the import/export code that I wrote is unambiguous, the case that @AlanCoding is talking about here cannot arise.

Reply:
I'm open to input but I think I am partial to the explicit `all` parameter and keywords. We could make it do something like:
```
- name: Export all resources
   tower_export:
   register: all_exported_data
```

And then have something like your code:
```
        # If no resource parameters are explicitly used, export everything.
        all_resources = all(module.params.get(resource) is None for resource in EXPORTABLE_RESOURCES)
```

But then we are back to our existing loop, just changing `module.params.get('all')` to `all_resources`.

I think I am also partial to the explicit:
```
- name: Export all credentials
  tower_export:
    credentials: 'all'
```

as opposed to :
```
- name: Export all credentials
  tower_export:
    credentials: ''
```

The second syntax looks like it could be confused with "nothing" vs an all. Anyone else have thoughts on this?

Reply:
Hmm, as a gut reaction I personally like the aesthetics of this

```yaml
- name: Export all resources
   tower_export:
   register: all_exported_data
```

But the behavior is "jumpy". If you have `credentials: "foo"` then it exports 1 thing, but take it away and it exports 1,000 things. If this syntax returned _nothing_ that would give a better impression of behavioral stability. I'm just saying it's arguable so I don't want to take a position.

I don't have a problem with the current "all" param. The validation statement I'm talking about would just be 1 conditional.

I also don't want to care about giving blank values. Ansible has an `omit` syntax. You could throw an error for blank values for clarity, but it doesn't bother me either way, it shouldn't be your problem to deal with.